autoload/db/adapter/redis.vim

```diff
@@ -7,7 +9,7 @@ function! db#adapter#redis#canonicalize(url) abort
 endfunction
 
 function! db#adapter#redis#interactive(url) abort
-  return ['redis-cli'] + db#url#as_argv(a:url, '-h ', '-p ', '', '', '-a ', '-n ')
+  return ['redis-cli'] + db#url#as_argv(a:url, '-h ', '-p ', '', '--user ', '-a ', '-n ')
```

Comment:
How backwards compatible is this? When I implemented this adapter, Redis didn't support usernames at all.

Reply:
In the current implementation the username portion would just be ignored.  When testing these are the scenarios and the outcome (before and after).

```vimscript
" Original
echo db#adapter#dispatch("redis://pass@local:1234/0", "interactive")
" ['redis-cli', '-h', 'local', '-p', '1234', '-n', '0']

echo db#adapter#dispatch("redis://:pass@local:1234/0", "interactive")
" ['redis-cli', '-h', 'local', '-p', '1234', '-a', 'pass', '-n', '0']

echo db#adapter#dispatch("redis://user:pass@local:1234/0", "interactive")
" ['redis-cli', '-h', 'local', '-p', '1234', '-a', 'pass', '-n', '0']

" Updated
echo db#adapter#dispatch("redis://pass@local:1234/0", "interactive")
" ['redis-cli', '-h', 'local', '-p', '1234', '--user', 'pass', '-n', '0']

echo db#adapter#dispatch("redis://:pass@local:1234/0", "interactive")
" ['redis-cli', '-h', 'local', '-p', '1234', '-a', 'pass', '-n', '0']

echo db#adapter#dispatch("redis://user:pass@local:1234/0", "interactive")
" ['redis-cli', '-h', 'local', '-p', '1234', '--user', 'user', '-a', 'pass', '-n', '0']
```

Depending on how people were trying to pass the password only solution, this would introduce an issue.  If they just left the user section empty but follow the url username/pass pattern it shouldn't cause a breaking change, but if they added any other arbitrary value in the username section of the url, this would not work.

What would be the preferred way to move forward?

Reply:
The docs explicitly tell users to use a dummy username with Redis:

https://github.com/tpope/vim-dadbod/blob/7888cb7164d69783d3dce4e0283decd26b82538b/doc/dadbod.txt#L190-L191

This matched the popular practice at the time. Providers like Heroku would use a meaningless username like `api`.

Another way to ask my question would be, how did everyone else handle this? Dadbod isn't the first project to retrofit usernames.

Reply:
:thinking: Let me see what I can find.

Reply:
Looks like Heroku uses `redis://:password@host` now. So I think this change is fine; just update the docs to mention an empty username rather than a dummy username.

Reply:
Just from some quick check these are a few references:
* I don't personally use heroku, but it appears that heroku uses an `h` for the username and around 2020 they [made an announcement](https://devcenter.heroku.com/changelog-items/1932) that they would start raising an error for users who passed `h` in for the username and would just move forward with a `<blank>` username as the recommended option.  (Although their [devcenter docs](https://devcenter.heroku.com/changelog-items/1932) still have the `h` username in them :question: )
* [This tool](https://shlink.io/documentation/advanced/using-redis/#supported-capabilities) was supporting an invalid URI (since username wasn't required at all) and they deprecated that form (granted they used `tls://` instead of `redis://` and doesn't exactly fit this situation.
* The redis entry in [Iana.org](https://www.iana.org/assignments/uri-schemes/prov/redis)  for the `redis://` format did call out the consideration of leaving the username blank to allow for the future use of a username (covered in the "Interoperability considerations" section).

For dadbob, the documentation didn't restrict the username, like heroku, to a common string (so no straight foward way to alert on valid or not valid), so that solution doesn't seem to apply.  It also didn't push for the empty string so I don't think that automatic integration is available.  That would mean that a solution to force the specific implementation would be in order.

My recommendation would be to add a new variable that can be set by users of the plugin to enable the user of the username (which would, of course, default to off to be backward compatible).  Something akin to `g:dadbod_redis_username` that can be set to enable the username flow.

If you wanted to do this as a temporary measure, I would also add an alert for users that are using a username and do not have the `dadbod_redis_username` variable set.  This alert should recommend they set the username to empty and also provide a timeline when the default will be to have username supported enabled.

Reply:
I'm okay if users have to update their URLs, as long as the solution isn't too obtuse and doesn't create incompatibilities elsewhere. From what you've gathered, it sounds like a blank username meets the requirements. So I think your implementation is good as is.